### PR TITLE
Disable debugging extension in ANGLE and refactor vertex array object extension initialization

### PR DESCRIPTION
**Pull request by anderco**
Crashes were reported when running debug builds on top of ANGLE. This PR disables said extension on that platform and refactors the vertex array object while at it.

/cc @kkaefer @brunoabinader

### CODE REVIEW

File: src/mbgl/gl/context.cpp [hunk 1]
```diff
@@ -108,6 +108,14 @@ Context::~Context() {
     }
 }
 
+template <typename Ex, typename Fn>
+std::unique_ptr<Ex> Context::initializeExtension(const Fn& loadExtension, std::function<bool()> blacklist) {
+    if (!blacklist())
+        return std::make_unique<Ex>(loadExtension);
```
Comment (kkaefer) [inline src/mbgl/gl/context.cpp]: Let's add braces around all blocks, even if they just consist of one statement.

File: src/mbgl/gl/context.cpp [hunk 2]
```diff
@@ -108,6 +108,14 @@ Context::~Context() {
     }
 }
 
+template <typename Ex, typename Fn>
+std::unique_ptr<Ex> Context::initializeExtension(const Fn& loadExtension, std::function<bool()> blacklist) {
```
Comment (kkaefer) [inline src/mbgl/gl/context.cpp]: `std::function<bool()>&&`

File: src/mbgl/gl/context.cpp [hunk 3]
```diff
@@ -124,10 +132,23 @@ void Context::initializeExtensions(const std::function<gl::ProcAddress(const cha
...
 
-        debugging = std::make_unique<extension::Debugging>(fn);
-        if (!disableVAOExtension) {
-            vertexArray = std::make_unique<extension::VertexArray>(fn);
-        }
+        const std::string renderer = reinterpret_cast<const char*>(MBGL_CHECK_ERROR(glGetString(GL_RENDERER)));
+        Log::Info(Event::General, "GPU Identifier: %s", renderer.c_str());
+
+        debugging = initializeExtension<extension::Debugging>(fn, [&renderer]() {
+            // Blacklist ANGLE since the debugging extension is causing crashes
```
Comment (kkaefer) [inline src/mbgl/gl/context.cpp]: While we're touching this code, let's stop using the word "blacklist" and replace it with something like "blocklist", or "stop list".

File: src/mbgl/gl/context.cpp [hunk 4]
```diff
@@ -108,6 +108,14 @@ Context::~Context() {
     }
 }
 
+template <typename Ex, typename Fn>
+std::unique_ptr<Ex> Context::initializeExtension(const Fn& loadExtension, std::function<bool()> blacklist) {
+    if (!blacklist())
+        return std::make_unique<Ex>(loadExtension);
+    else
+        return {};
+}
```
Comment (kkaefer) [inline src/mbgl/gl/context.cpp]: I'm not sure I see the need for this function. Can we instead use a simpler

```cpp
if (<code to determine whether it the extension is available>) {
    extension = std::make_unique<Extension>(loadExtension);
}
```

File: src/mbgl/gl/context.cpp [hunk 5]
```diff
@@ -124,10 +132,23 @@ void Context::initializeExtensions(const std::function<gl::ProcAddress(const cha
...
+            // Blacklist ANGLE since the debugging extension is causing crashes
+            return renderer.find("ANGLE") != std::string::npos;
+        });
+        vertexArray = initializeExtension<extension::VertexArray>(fn, [&renderer]() {
+            // Blacklist Adreno 2xx, 3xx as it crashes on glBuffer(Sub)Data
+            // Blacklist ARM Mali-T720 (in some MT8163 chipsets) as it crashes on glBindVertexArray
+            // Blacklist ANGLE as the combination of Qt + Windows + ANGLE leads to crashes
+            return renderer.find("Adreno (TM) 2") != std::string::npos
+                || renderer.find("Adreno (TM) 3") != std::string::npos
+                || renderer.find("ANGLE") != std::string::npos
```
Comment (kkaefer) [inline src/mbgl/gl/context.cpp]: Why did you add ANGLE to the VertexArray exclusion as well?

File: src/mbgl/gl/context.cpp (hunk 3 above)
Comment (anderco) [reply src/mbgl/gl/context.cpp]: Sure, will do.

File: src/mbgl/gl/context.cpp (hunk 5 above)
Comment (anderco) [reply src/mbgl/gl/context.cpp]: I tried to capture that in the first commit's message: https://github.com/mapbox/mapbox-gl-native/pull/12754/commits/d6180d1cbfbfcf30803b2792e54ab3df8dffdf99

The idea is to unify the \#ifdef disabling the extension on Windows with the blocklist. There's an assumption that ANGLE is only being used on Windows, though.

Comment (anderco) [reply src/mbgl/gl/context.cpp]: I removed the assumption on ANGLE only being used on windows. The extensions are only disabled now if ANGLE is running on Direct3D.

File: src/mbgl/gl/context.cpp [hunk 6]
```diff
@@ -124,10 +124,27 @@ void Context::initializeExtensions(const std::function<gl::ProcAddress(const cha
...
 
-        debugging = std::make_unique<extension::Debugging>(fn);
-        if (!disableVAOExtension) {
-            vertexArray = std::make_unique<extension::VertexArray>(fn);
+        const std::string renderer = reinterpret_cast<const char*>(MBGL_CHECK_ERROR(glGetString(GL_RENDERER)));
+        Log::Info(Event::General, "GPU Identifier: %s", renderer.c_str());
+
+        // Block ANGLE on Direct3D since the debugging extension is causing crashes
+        if (renderer.find("ANGLE") == std::string::npos
+            || renderer.find("Direct3D") == std::string::npos) {
```
Comment (kkaefer) [inline src/mbgl/gl/context.cpp]: While this may be the correct logic (disable Debugging for Angle on Direct3D), it's a bit confusing to read this code. I think

```cpp
!(renderer.find("ANGLE") != std::string::npos && renderer.find("Direct3D") != std::string::npos)
```

may be easier to read.

File: src/mbgl/gl/context.cpp [hunk 7]
```diff
@@ -124,10 +124,27 @@ void Context::initializeExtensions(const std::function<gl::ProcAddress(const cha
             return nullptr;
         };
 
-        debugging = std::make_unique<extension::Debugging>(fn);
-        if (!disableVAOExtension) {
```
Comment (kkaefer) [inline src/mbgl/gl/context.cpp]: We're using `disableVAOExtension` in https://github.com/mapbox/mapbox-gl-native/blob/8ec92b72473305cbf14c31b185fd5634ce670837/test/map/map.test.cpp#L450-L459
so it needs to be re-added to the exclusion logic below

File: src/mbgl/gl/context.cpp (hunk 6 above)
Comment (anderco) [reply src/mbgl/gl/context.cpp]: :+1: I actually wrote that first and then started wondering if the double negation made it more confusing. It seems it doesn't. I'll change it.
